packages/common/src/store/pages/saved-page/reducer.ts

```diff
@@ -21,19 +29,106 @@ import { signOut } from 'store/sign-out/slice'
 import { ActionsMap } from 'utils/reducer'
 
 import { PREFIX as tracksPrefix } from './lineups/tracks/actions'
-import { LibraryCategory, SavedPageState } from './types'
+import { LibraryCategory, LibraryCategoryType, SavedPageState } from './types'
 
 const initialState = {
   // id => uid
-  localSaves: {},
-  saves: [],
+  localTrackFavorites: {},
+  localTrackPurchases: {},
+  localTrackReposts: {},
+  localAlbumFavorites: [],
+  localAlbumPurchases: [],
+  localAlbumReposts: [],
+  localRemovedAlbumFavorites: [],
+  localRemovedAlbumReposts: [],
+  localPlaylistFavorites: [],
+  localPlaylistReposts: [],
+  localPlaylistPurchases: [],
+  localRemovedPlaylistFavorites: [],
+  localRemovedPlaylistResposts: [],
+  trackSaves: [],
   initialFetch: false,
   hasReachedEnd: false,
   fetchingMore: false,
   tracks: initialLineupState,
   selectedCategory: LibraryCategory.Favorite
 } as SavedPageState
 
+/** Utility to get the name of key in which locally added or removed collections are stored in SavedPageState.
+ * For example, playlists that are favorited/unfavorited locally go into `localPlaylistFavorites`/`localPlaylistRemovedFavorites`.
+ */
+const getLocalCollectionStateKeys = ({
```

Comment:
Just a thought: If all of these follow similar patterns, you could just structure this as nested objects based on that pattern, i.e. `state.local[kind][category][mutation]` where
 * `kind` is Track, Playlist, Album and an enum
 * `category` is your existing category
 * `mutation` is 'added' or 'removed'

That would still allow you to strictly define the type in a way that doesn't require building a string for indexing.

This could also let you pare down the actions to a single set where you pass the kind/category/mutation values instead of an action for each permutation of all the values together.

edit: I can see below why having it all flattened makes the merging logic easier (spread the state and modify just the set of add/remove values for the kind/category you are changing)

Reply:
ah i like this idea! will ponder